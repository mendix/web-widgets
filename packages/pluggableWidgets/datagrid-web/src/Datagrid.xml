<?xml version="1.0" encoding="utf-8" ?>
<widget id="com.mendix.widget.web.datagrid.Datagrid" pluginWidget="true" offlineCapable="true" xmlns="http://www.mendix.com/widget/1.0/" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xsi:schemaLocation="http://www.mendix.com/widget/1.0/ ../node_modules/mendix/custom_widget.xsd">
    <name>Data grid 2</name>
    <description />
    <studioProCategory>Data containers</studioProCategory>
    <studioCategory>Data Containers</studioCategory>
    <helpUrl>https://docs.mendix.com/appstore/modules/data-grid-2</helpUrl>
    <properties>
        <propertyGroup caption="General">
            <propertyGroup caption="General">
                <property key="advanced" type="boolean" defaultValue="false">
                    <caption>Enable advanced options</caption>
                    <description />
                </property>
                <property key="datasource" type="datasource" isList="true">
                    <caption>Data source</caption>
                    <description />
                </property>
                <property key="refreshInterval" type="integer" defaultValue="0">
                    <caption>Refresh time (in seconds)</caption>
                    <description />
                </property>
                <property key="itemSelection" type="selection" dataSource="datasource">
                    <caption>Selection</caption>
                    <description />
                    <selectionTypes>
                        <selectionType name="None" />
                        <selectionType name="Single" />
                        <selectionType name="Multi" />
                    </selectionTypes>
                </property>
                <property key="itemSelectionMethod" type="enumeration" defaultValue="checkbox">
                    <caption>Selection method</caption>
                    <description />
                    <enumerationValues>
                        <enumerationValue key="checkbox">Checkbox</enumerationValue>
                        <enumerationValue key="rowClick">Row click</enumerationValue>
                    </enumerationValues>
                </property>
                <property key="itemSelectionMode" type="enumeration" defaultValue="clear">
                    <caption>Toggle on click</caption>
                    <description>Defines item selection behavior.</description>
                    <enumerationValues>
                        <enumerationValue key="toggle">Yes</enumerationValue>
                        <enumerationValue key="clear">No</enumerationValue>
                    </enumerationValues>
                </property>
                <property key="showSelectAllToggle" type="boolean" defaultValue="true">
                    <caption>Show (un)check all toggle</caption>
                    <description>Show a checkbox in the grid header to check or uncheck multiple items.</description>
                </property>
            </propertyGroup>
            <propertyGroup caption="Columns">
                <property key="columns" type="object" isList="true">
                    <caption>Columns</caption>
                    <description />
                    <properties>
                        <propertyGroup caption="General">
                            <property key="showContentAs" type="enumeration" defaultValue="attribute">
                                <caption>Show</caption>
                                <description />
                                <enumerationValues>
                                    <enumerationValue key="attribute">Attribute</enumerationValue>
                                    <enumerationValue key="dynamicText">Dynamic text</enumerationValue>
                                    <enumerationValue key="customContent">Custom content</enumerationValue>
                                </enumerationValues>
                            </property>
                            <property key="attribute" type="attribute" dataSource="../datasource" required="false">
                                <caption>Attribute</caption>
                                <description>Attribute is required if the column can be sorted or filtered</description>
                                <attributeTypes>
                                    <attributeType name="String" />
                                    <attributeType name="AutoNumber" />
                                    <attributeType name="Boolean" />
                                    <attributeType name="DateTime" />
                                    <attributeType name="Decimal" />
                                    <attributeType name="Enum" />
                                    <attributeType name="Integer" />
                                    <attributeType name="Long" />
                                </attributeTypes>
                            </property>
                            <property key="content" type="widgets" dataSource="../datasource" required="false">
                                <caption>Custom content</caption>
                                <description />
                            </property>
                            <property key="dynamicText" type="textTemplate" dataSource="../datasource" required="false">
                                <caption>Dynamic text</caption>
                                <description />
                            </property>
                            <property key="header" type="textTemplate" required="false">
                                <caption>Caption</caption>
                                <description />
                            </property>
                            <property key="tooltip" type="textTemplate" required="false" dataSource="../datasource">
                                <caption>Tooltip</caption>
                                <description />
                            </property>
                            <property key="filter" type="widgets" required="false">
                                <caption>Filter</caption>
                                <description />
                            </property>
                            <property key="filterAssociation" type="association" selectableObjects="filterAssociationOptions" required="false" dataSource="../datasource">
                                <caption>Reference</caption>
                                <description>Set the reference to enable filtering over association with the Drop-down filter widget.</description>
                                <associationTypes>
                                    <associationType name="Reference" />
                                    <associationType name="ReferenceSet" />
                                </associationTypes>
                            </property>
                            <property key="filterAssociationOptions" type="datasource" isList="true" required="false">
                                <caption>Data source</caption>
                                <description>The options to show in the Drop-down filter widget.</description>
                            </property>
                            <property key="filterAssociationOptionLabel" type="expression" dataSource="filterAssociationOptions" required="false">
                                <caption>Option caption</caption>
                                <description />
                                <returnType type="String" />
                            </property>
                            <property key="visible" type="expression" required="true" defaultValue="true">
                                <caption>Visible</caption>
                                <description />
                                <returnType type="Boolean" />
                            </property>
                        </propertyGroup>
                        <propertyGroup caption="Column capabilities">
                            <property key="sortable" type="boolean" defaultValue="true">
                                <caption>Can sort</caption>
                                <description />
                            </property>
                            <property key="resizable" type="boolean" defaultValue="true">
                                <caption>Can resize</caption>
                                <description />
                            </property>
                            <property key="draggable" type="boolean" defaultValue="true">
                                <caption>Can reorder</caption>
                                <description />
                            </property>
                            <property key="hidable" type="enumeration" defaultValue="yes">
                                <caption>Can hide</caption>
                                <description />
                                <enumerationValues>
                                    <enumerationValue key="yes">Yes</enumerationValue>
                                    <enumerationValue key="hidden">Yes, hidden by default</enumerationValue>
                                    <enumerationValue key="no">No</enumerationValue>
                                </enumerationValues>
                            </property>
                            <property key="allowEventPropagation" type="boolean" defaultValue="true">
                                <caption>Allow row events</caption>
                                <description>If set to yes, then all default events on the row, such as "on click" or selection, will be triggered when the user interacts with custom content.</description>
                            </property>
                        </propertyGroup>
                        <propertyGroup caption="Appearance">
                            <property key="width" type="enumeration" defaultValue="autoFill">
                                <caption>Column width</caption>
                                <description />
                                <enumerationValues>
                                    <enumerationValue key="autoFill">Auto-fill</enumerationValue>
                                    <enumerationValue key="autoFit">Auto-fit content</enumerationValue>
                                    <enumerationValue key="manual">Manual</enumerationValue>
                                </enumerationValues>
                            </property>
                            <property key="minWidth" type="enumeration" defaultValue="auto">
                                <caption>Min width</caption>
                                <description />
                                <enumerationValues>
                                    <enumerationValue key="auto">Auto</enumerationValue>
                                    <enumerationValue key="minContent">Set by content</enumerationValue>
                                    <enumerationValue key="manual">Manual</enumerationValue>
                                </enumerationValues>
                            </property>
                            <property key="minWidthLimit" type="integer" defaultValue="100">
                                <caption>Min width value (px)</caption>
                                <description />
                            </property>
                            <property key="size" type="integer" defaultValue="1">
                                <caption>Column size</caption>
                                <description />
                            </property>
                            <property key="alignment" type="enumeration" defaultValue="left">
                                <caption>Alignment</caption>
                                <description />
                                <enumerationValues>
                                    <enumerationValue key="left">Left</enumerationValue>
                                    <enumerationValue key="center">Center</enumerationValue>
                                    <enumerationValue key="right">Right</enumerationValue>
                                </enumerationValues>
                            </property>
                            <property key="columnClass" type="expression" required="false" dataSource="../datasource">
                                <caption>Dynamic cell class</caption>
                                <description />
                                <returnType type="String" />
                            </property>
                            <property key="wrapText" type="boolean" defaultValue="false">
                                <caption>Wrap text</caption>
                                <description />
                            </property>
                        </propertyGroup>
                    </properties>
                </property>
                <property key="columnsFilterable" type="boolean" defaultValue="true">
                    <caption>Show column filters</caption>
                    <description />
                </property>
            </propertyGroup>
            <propertyGroup caption="Rows">
                <property key="pageSize" type="integer" defaultValue="20">
                    <caption>Page size</caption>
                    <description />
                </property>
                <property key="pagination" type="enumeration" defaultValue="buttons">
                    <caption>Pagination</caption>
                    <description />
                    <enumerationValues>
                        <enumerationValue key="buttons">Paging buttons</enumerationValue>
                        <enumerationValue key="virtualScrolling">Virtual scrolling</enumerationValue>
                        <enumerationValue key="loadMore">Load more</enumerationValue>
                    </enumerationValues>
                </property>
                <property key="pagingPosition" type="enumeration" defaultValue="bottom">
                    <caption>Position of paging buttons</caption>
                    <description />
                    <enumerationValues>
                        <enumerationValue key="bottom">Below grid</enumerationValue>
                        <enumerationValue key="top">Above grid</enumerationValue>
                        <enumerationValue key="both">Both</enumerationValue>
                    </enumerationValues>
                </property>
                <property key="showPagingButtons" type="enumeration" defaultValue="always">
                    <caption>Show paging buttons</caption>
                    <description />
                    <enumerationValues>
                        <enumerationValue key="always">Always</enumerationValue>
                        <enumerationValue key="auto">Auto</enumerationValue>
                    </enumerationValues>
                </property>
                <property key="loadMoreButtonCaption" type="textTemplate" required="false">
                    <caption>Load more caption</caption>
                    <description />
                    <translations>
                        <translation lang="en_US">Load More</translation>
                    </translations>
                </property>
                <property key="showEmptyPlaceholder" type="enumeration" defaultValue="none">
                    <caption>Empty list message</caption>
                    <description />
                    <enumerationValues>
                        <enumerationValue key="none">None</enumerationValue>
                        <enumerationValue key="custom">Custom</enumerationValue>
                    </enumerationValues>
                </property>
                <property key="emptyPlaceholder" type="widgets" required="false">
                    <caption>Empty placeholder</caption>
                    <description />
                </property>
                <property key="rowClass" type="expression" required="false" dataSource="datasource">
                    <caption>Dynamic row class</caption>
                    <description />
                    <returnType type="String" />
                </property>
            </propertyGroup>
            <propertyGroup caption="Events">
                <property key="onClickTrigger" type="enumeration" defaultValue="single">
                    <caption>On click trigger</caption>
                    <description />
                    <enumerationValues>
                        <enumerationValue key="single">Single click</enumerationValue>
                        <enumerationValue key="double">Double click</enumerationValue>
                    </enumerationValues>
                </property>
                <property key="onClick" type="action" required="false" dataSource="datasource">
                    <caption>On click action</caption>
                    <description />
                </property>
                <property key="onSelectionChange" type="action" required="false">
                    <caption>On selection change</caption>
                    <description />
                </property>
            </propertyGroup>
        </propertyGroup>
        <propertyGroup caption="Personalization">
            <propertyGroup caption="Column capabilities">
                <property key="columnsSortable" type="boolean" defaultValue="true">
                    <caption>Sorting</caption>
                    <description>Enable sorting for all columns unless specified otherwise in the column setting</description>
                </property>
                <property key="columnsResizable" type="boolean" defaultValue="true">
                    <caption>Resizing</caption>
                    <description>Enable resizing for all columns unless specified otherwise in the column setting</description>
                </property>
                <property key="columnsDraggable" type="boolean" defaultValue="true">
                    <caption>Reordering</caption>
                    <description>Enable reordering for all columns unless specified otherwise in the column setting</description>
                </property>
                <property key="columnsHidable" type="boolean" defaultValue="true">
                    <caption>Hiding</caption>
                    <description>Enable hiding for all columns unless specified otherwise in the column setting</description>
                </property>
            </propertyGroup>
            <propertyGroup caption="Configuration">
                <property key="configurationStorageType" type="enumeration" defaultValue="attribute">
                    <caption>Store configuration in</caption>
                    <description>When Browser local storage is selected, the configuration is scoped to a browser profile. This configuration is not tied to a Mendix user.</description>
                    <enumerationValues>
                        <enumerationValue key="attribute">Attribute</enumerationValue>
                        <enumerationValue key="localStorage">Browser local storage</enumerationValue>
                    </enumerationValues>
                </property>
                <property key="configurationAttribute" type="attribute" required="false" onChange="onConfigurationChange">
                    <caption>Attribute</caption>
                    <description>Attribute containing the personalized configuration of the capabilities. This configuration is automatically stored and loaded. The attribute requires Unlimited String.</description>
                    <attributeTypes>
                        <attributeType name="String" />
                    </attributeTypes>
                </property>
                <property key="onConfigurationChange" type="action" required="false">
                    <caption>On change</caption>
                    <description />
                </property>
            </propertyGroup>
        </propertyGroup>
        <propertyGroup caption="Grid wide filtering">
            <propertyGroup caption="Grid wide filtering">
                <property key="filterList" type="object" isList="true" required="false">
                    <caption>Filters</caption>
                    <description>The list of attributes is used by the filter widgets that are placed in the placeholder above the data grid. This enables filtering across multiple attributes or columns instead of limiting to a single column.</description>
                    <properties>
                        <propertyGroup caption="General">
                            <property key="filter" type="attribute" dataSource="../datasource">
                                <caption>Filter attribute</caption>
                                <description />
                                <attributeTypes>
                                    <attributeType name="String" />
                                    <attributeType name="AutoNumber" />
                                    <attributeType name="Boolean" />
                                    <attributeType name="DateTime" />
                                    <attributeType name="Decimal" />
                                    <attributeType name="Enum" />
                                    <attributeType name="Integer" />
                                    <attributeType name="Long" />
                                </attributeTypes>
                            </property>
                        </propertyGroup>
                    </properties>
                </property>
                <property key="filtersPlaceholder" type="widgets" required="false">
                    <caption>Filters placeholder</caption>
                    <description />
                </property>
            </propertyGroup>
        </propertyGroup>
        <propertyGroup caption="Accessibility">
            <propertyGroup caption="Aria labels">
                <property key="filterSectionTitle" type="textTemplate" required="false">
                    <caption>Filter section</caption>
                    <description>Assistive technology will read this upon reaching a filtering or sorting section.</description>
                </property>
                <property key="exportDialogLabel" type="textTemplate" required="false">
                    <caption>Export progress</caption>
                    <description>Assistive technology will read this upon reaching a export dialog.</description>
                    <translations>
                        <translation lang="en_US">Export progress</translation>
                    </translations>
                </property>
                <property key="cancelExportLabel" type="textTemplate" required="false">
                    <caption>Cancel data export</caption>
                    <description>Assistive technology will read this upon reaching a cancel button.</description>
                    <translations>
                        <translation lang="en_US">Cancel data export</translation>
                    </translations>
                </property>
                <property key="selectRowLabel" type="textTemplate" required="false">
                    <caption>Select row</caption>
                    <description>If selection is enabled, assistive technology will read this upon reaching a checkbox.</description>
                    <translations>
                        <translation lang="en_US">Select row</translation>
                    </translations>
                </property>
            </propertyGroup>
        </propertyGroup>
    </properties>
</widget>
