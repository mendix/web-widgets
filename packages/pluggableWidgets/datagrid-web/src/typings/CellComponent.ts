import { ObjectItem } from "mendix";
import { ReactElement, ReactNode } from "react";
import { GridColumn } from "./GridColumn";
import { ElementProps } from "@mendix/widget-plugin-grid/event-switch/base";

export interface EventsController {
    getProps(item: ObjectItem): ElementProps<HTMLDivElement>;
}

export interface CellComponentProps<C extends GridColumn> {
    children?: ReactNode;
    className?: string;
    column: C;
    item: ObjectItem;
    key?: string | number;
    rowIndex: number;
    columnIndex?: number;
    clickable?: boolean;
    preview?: boolean;
    eventsController: EventsController;
}

export type CellComponent<C extends GridColumn> = (props: CellComponentProps<C>) => ReactElement;
