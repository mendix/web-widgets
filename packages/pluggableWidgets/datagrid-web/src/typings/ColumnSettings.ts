import { ColumnId } from "./GridColumn";

export interface ColumnSettings {
    columnId: ColumnId;
    size: number | undefined;
    hidden: boolean;
    filterSettings: unknown;
}
