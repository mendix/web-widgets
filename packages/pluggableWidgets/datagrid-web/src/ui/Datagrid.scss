@use "./ExportAlert.scss";
@use "./PseudoModal.scss";
@use "./ExportProgress.scss";

.widget-datagrid {
    position: relative;
    .th.widget-datagrid-col-select,
    .td.widget-datagrid-col-select {
        align-items: center;
    }
}

.widget-datagrid-exporting {
    .widget-datagrid-top-bar,
    .widget-datagrid-header,
    .widget-datagrid-content,
    .widget-datagrid-footer {
        visibility: hidden;

        * {
            transition: unset;
        }
    }
}
.widget-datagrid-load-more {
    display: block !important;
    margin: 0 auto;
}
