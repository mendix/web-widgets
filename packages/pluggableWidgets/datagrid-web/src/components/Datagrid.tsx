import {
    FilterState,
    FilterType,
    useFilterContext,
    useMultipleFiltering,
    readInitFilterValues
} from "@mendix/widget-plugin-filtering";
import { useCreateSelectionContextValue, useSelectionHelper } from "@mendix/widget-plugin-grid/selection";
import { useGridSelectionProps } from "@mendix/widget-plugin-grid/selection/useGridSelectionProps";
import { generateUUID } from "@mendix/widget-plugin-platform/framework/generate-uuid";
import { FilterCondition } from "mendix/filters";
import { and } from "mendix/filters/builders";
import { ReactElement, ReactNode, createElement, useCallback, useRef, useState } from "react";
import { DatagridContainerProps } from "../../typings/DatagridProps";
import { Cell } from "./Cell";
import { Widget } from "./Widget";
import { WidgetHeaderContext } from "./WidgetHeaderContext";
import { getColumnAssociationProps } from "../features/column";
import { UpdateDataSourceFn, useDG2ExportApi } from "../features/export";
import { Model } from "../features/model/base";
import { useStoreMap, useUnit } from "effector-react";
import { useViewModel } from "../features/model/use-view-model";
import { Actions } from "../typings/GridModel";

type Props = DatagridContainerProps & {
    model: Model;
    actions: Actions;
};

export default function Datagrid(props: Props): ReactElement {
    const id = useRef(`DataGrid${generateUUID()}`);
    const multipleFilteringState = useMultipleFiltering();
    const { FilterContext } = useFilterContext();
    const { model, actions } = props;

    const viewModel = useViewModel(model);

    const exportColumns = useStoreMap(props.model.visible, visible =>
        visible.map(col => props.columns[col.columnNumber])
    );

    const gridFilter = useUnit(model.filter);

    const [{ items, exporting, processedRows }, { abort }] = useDG2ExportApi({
        columns: exportColumns,
        hasMoreItems: props.datasource.hasMoreItems || false,
        items: props.datasource.items,
        name: props.name,
        offset: props.datasource.offset,
        limit: props.datasource.limit,
        updateDataSource: useCallback<UpdateDataSourceFn>(
            ({ offset, limit, reload }) => {
                if (offset != null) {
                    props.datasource?.setOffset(offset);
                }

                if (limit != null) {
                    props.datasource?.setLimit(limit);
                }

                if (reload) {
                    props.datasource.reload();
                }
            },
            [props.datasource]
        )
    });

    // TODO: Rewrite this logic with single useReducer (or write
    // custom hook that will use useReducer)
    // eslint-disable-next-line react-hooks/rules-of-hooks
    const customFiltersState = props.columns.map(() => useState<FilterState>());

    const filters = customFiltersState
        .map(([customFilter]) => customFilter?.getFilterCondition?.())
        .filter((filter): filter is FilterCondition => filter !== undefined)
        .concat(
            // Concatenating multiple filter state
            Object.keys(multipleFilteringState)
                .map((key: FilterType) => multipleFilteringState[key][0]?.getFilterCondition())
                .filter((filter): filter is FilterCondition => filter !== undefined)
        );

    if (filters.length > 0) {
        actions.setFilter(filters.length > 1 ? and(...filters) : filters[0]);
    } else {
        actions.setFilter(undefined);
    }

    const selectionHelper = useSelectionHelper(
        props.itemSelection,
        props.datasource,
        props.onSelectionChange,
        props.pageSize
    );
    const selectionContextValue = useCreateSelectionContextValue(selectionHelper);
    const selectionProps = useGridSelectionProps({
        selection: props.itemSelection,
        selectionMethod: props.itemSelectionMethod,
        helper: selectionHelper,
        showSelectAllToggle: props.showSelectAllToggle
    });

    return (
        <Widget
            className={props.class}
            CellComponent={Cell}
            columnsDraggable={props.columnsDraggable}
            columnsFilterable={props.columnsFilterable}
            columnsHidable={props.columnsHidable}
            columnsResizable={props.columnsResizable}
            columnsSortable={props.columnsSortable}
            data={items}
            emptyPlaceholderRenderer={useCallback(
                (renderWrapper: (children: ReactNode) => ReactElement) =>
                    props.showEmptyPlaceholder === "custom" ? renderWrapper(props.emptyPlaceholder) : <div />,
                [props.emptyPlaceholder, props.showEmptyPlaceholder]
            )}
            filterRenderer={useCallback(
                (renderWrapper, column) => {
                    const rawColumn = props.columns[column.columnNumber];
                    const { attribute, filter } = rawColumn;
                    const associationProps = getColumnAssociationProps(rawColumn);

                    const initialFilters = readInitFilterValues(attribute, gridFilter);

                    if (!attribute && !associationProps) {
                        return renderWrapper(filter);
                    }

                    return renderWrapper(
                        <FilterContext.Provider
                            value={{
                                filterDispatcher: state => actions.setColumnFilter([column.columnId, state]),
                                singleAttribute: attribute,
                                singleInitialFilter: initialFilters,
                                associationProperties: associationProps
                            }}
                        >
                            {filter}
                        </FilterContext.Provider>
                    );
                },
                // eslint-disable-next-line react-hooks/exhaustive-deps
                [FilterContext, customFiltersState, props.columns]
            )}
            headerTitle={props.filterSectionTitle?.value}
            headerContent={
                props.filtersPlaceholder && (
                    <WidgetHeaderContext
                        filterList={props.filterList}
                        viewStateFilters={undefined}
                        selectionContextValue={selectionContextValue}
                        state={multipleFilteringState}
                    >
                        {props.filtersPlaceholder}
                    </WidgetHeaderContext>
                )
            }
            hasMoreItems={props.datasource.hasMoreItems ?? false}
            headerWrapperRenderer={useCallback((_columnIndex: number, header: ReactElement) => header, [])}
            id={id.current}
            numberOfItems={props.datasource.totalCount}
            onExportCancel={abort}
            pageSize={props.pageSize}
            paging={props.pagination === "buttons"}
            pagingPosition={props.pagingPosition}
            rowClass={useCallback((value: any) => props.rowClass?.get(value)?.value ?? "", [props.rowClass])}
            styles={props.style}
            rowAction={props.onClick}
            selectionProps={selectionProps}
            selectionStatus={selectionHelper?.type === "Multi" ? selectionHelper.selectionStatus : "unknown"}
            exporting={exporting}
            processedRows={processedRows}
            exportDialogLabel={props.exportDialogLabel?.value}
            cancelExportLabel={props.cancelExportLabel?.value}
            selectRowLabel={props.selectRowLabel?.value}
            actions={props.actions}
            model={viewModel}
        />
    );
}
