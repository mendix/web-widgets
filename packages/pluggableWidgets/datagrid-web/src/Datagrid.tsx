import { createElement, ReactElement, ReactNode, useCallback, useEffect, useMemo, useRef, useState } from "react";
import { ColumnsType, DatagridContainerProps } from "../typings/DatagridProps";
import { FilterCondition } from "mendix/filters";
import { and } from "mendix/filters/builders";
import { Table, TableColumn, SortProperty } from "./components/Table";
import {
    FilterFunction,
    FilterType,
    generateUUID,
    useFilterContext,
    useMultipleFiltering
} from "@mendix/pluggable-widgets-commons/components/web";
import {
    getGlobalSelectionContext,
    isAvailable,
    useCreateSelectionContextValue,
    useSelectionHelper
} from "@mendix/pluggable-widgets-commons";
import { extractFilters } from "./features/filters";
import { useCellRenderer } from "./features/cell";
import { getColumnAssociationProps, isSortable } from "./features/column";
import { selectionSettings, useOnSelectProps } from "./features/selection";
import "./ui/Datagrid.scss";

export default function Datagrid(props: DatagridContainerProps): ReactElement {
    const id = useRef(`DataGrid${generateUUID()}`);

    const [sortParameters, setSortParameters] = useState<SortProperty | undefined>(undefined);
    const isInfiniteLoad = props.pagination === "virtualScrolling";
    const currentPage = isInfiniteLoad
        ? props.datasource.limit / props.pageSize
        : props.datasource.offset / props.pageSize;
    const viewStateFilters = useRef<FilterCondition | undefined>(undefined);
    const [filtered, setFiltered] = useState(false);
    const multipleFilteringState = useMultipleFiltering();
    const { FilterContext } = useFilterContext();
    const SelectionContext = getGlobalSelectionContext();
    const cellRenderer = useCellRenderer({ columns: props.columns, onClick: props.onClick });

    useEffect(() => {
        props.datasource.requestTotalCount(!isInfiniteLoad);
        if (props.datasource.limit === Number.POSITIVE_INFINITY) {
            props.datasource.setLimit(props.pageSize);
        }
    }, [props.datasource, props.pageSize]);

    useEffect(() => {
        if (props.datasource.filter && !filtered && !viewStateFilters.current) {
            viewStateFilters.current = props.datasource.filter;
        }
    }, [props.datasource, props.configurationAttribute, filtered]);

    useEffect(() => {
        if (props.refreshInterval > 0) {
            setTimeout(() => {
                props.datasource.reload();
            }, props.refreshInterval * 1000);
        }
    }, [props.datasource, props.refreshInterval]);

    const setPage = useCallback(
        (computePage: (prevPage: number) => number) => {
            const newPage = computePage(currentPage);
            if (isInfiniteLoad) {
                props.datasource.setLimit(newPage * props.pageSize);
            } else {
                props.datasource.setOffset(newPage * props.pageSize);
            }
        },
        [props.datasource, props.pageSize, isInfiniteLoad, currentPage]
    );

    // TODO: Rewrite this logic with single useReducer (or write
    // custom hook that will use useReducer)
    // eslint-disable-next-line react-hooks/rules-of-hooks
    const customFiltersState = props.columns.map(() => useState<FilterFunction>());

    const filters = customFiltersState
        .map(([customFilter]) => customFilter?.getFilterCondition?.())
        .filter((filter): filter is FilterCondition => filter !== undefined)
        .concat(
            // Concatenating multiple filter state
            Object.keys(multipleFilteringState)
                .map((key: FilterType) => multipleFilteringState[key][0]?.getFilterCondition())
                .filter((filter): filter is FilterCondition => filter !== undefined)
        );

    if (filters.length > 0) {
        props.datasource.setFilter(filters.length > 1 ? and(...filters) : filters[0]);
    } else if (filtered) {
        props.datasource.setFilter(undefined);
    } else {
        props.datasource.setFilter(viewStateFilters.current);
    }

    if (sortParameters) {
        props.datasource.setSortOrder([
            [props.columns[sortParameters.columnIndex].attribute!.id, sortParameters.desc ? "desc" : "asc"]
        ]);
    } else {
        props.datasource.setSortOrder(undefined);
    }

    const columns = useMemo(() => transformColumnProps(props.columns), [props.columns]);

    /**
     * Multiple filtering properties
     */
    const filterList = useMemo(
        () => props.filterList.reduce((filters, { filter }) => ({ ...filters, [filter.id]: filter }), {}),
        [props.filterList]
    );
    const multipleInitialFilters = useMemo(
        () =>
            props.filterList.reduce(
                (filters, { filter }) => ({
                    ...filters,
                    [filter.id]: extractFilters(filter, viewStateFilters.current)
                }),
                {}
            ),
        [props.filterList]
    );

    const selection = useSelectionHelper(props.itemSelection, props.datasource, props.onSelectionChange);
    const selectActionProps = useOnSelectProps(selection);
    const { selectionStatus, selectionMethod } = selectionSettings(props, selection);

    const selectionContextValue = useCreateSelectionContextValue(selection);

    return (
        <Table
            selectionStatus={selectionStatus}
            selectionMethod={selectionMethod}
            cellRenderer={cellRenderer}
            className={props.class}
            columns={columns}
            columnsDraggable={props.columnsDraggable}
            columnsFilterable={props.columnsFilterable}
            columnsHidable={props.columnsHidable}
            columnsResizable={props.columnsResizable}
            columnsSortable={props.columnsSortable}
            data={props.datasource.items ?? []}
            emptyPlaceholderRenderer={useCallback(
                (renderWrapper: (children: ReactNode) => ReactElement) =>
                    props.showEmptyPlaceholder === "custom" ? renderWrapper(props.emptyPlaceholder) : <div />,
                [props.emptyPlaceholder, props.showEmptyPlaceholder]
            )}
            filterRenderer={useCallback(
                (renderWrapper, columnIndex) => {
                    const column = props.columns[columnIndex];
                    const { attribute, filter } = column;
                    const associationProps = getColumnAssociationProps(column);
                    const [, filterDispatcher] = customFiltersState[columnIndex];
                    const initialFilters = extractFilters(attribute, viewStateFilters.current);

                    if (!attribute && !associationProps) {
                        return renderWrapper(filter);
                    }

                    return renderWrapper(
                        <FilterContext.Provider
                            value={{
                                filterDispatcher: prev => {
                                    setFiltered(true);
                                    filterDispatcher(prev);
                                    return prev;
                                },
                                singleAttribute: attribute,
                                singleInitialFilter: initialFilters,
                                associationProperties: associationProps
                            }}
                        >
                            {filter}
                        </FilterContext.Provider>
                    );
                },
                [FilterContext, customFiltersState, props.columns]
            )}
            hasMoreItems={props.datasource.hasMoreItems ?? false}
            headerWrapperRenderer={useCallback((_columnIndex: number, header: ReactElement) => header, [])}
            gridHeaderWidgets={useMemo(
                () => (
                    <FilterContext.Provider
                        value={{
                            filterDispatcher: prev => {
                                if (prev.filterType) {
                                    const [, filterDispatcher] = multipleFilteringState[prev.filterType];
                                    filterDispatcher(prev);
                                    setFiltered(true);
                                }
                                return prev;
                            },
                            multipleAttributes: filterList,
                            multipleInitialFilters
                        }}
                    >
                        <SelectionContext.Provider value={selectionContextValue}>
                            {props.filtersPlaceholder}
                        </SelectionContext.Provider>
                    </FilterContext.Provider>
                ),
                [FilterContext, filterList, multipleInitialFilters, props.filtersPlaceholder, multipleFilteringState]
            )}
            gridHeaderTitle={props.filterSectionTitle?.value}
            id={id.current}
            numberOfItems={props.datasource.totalCount}
            page={currentPage}
            pageSize={props.pageSize}
            paging={props.pagination === "buttons"}
            pagingPosition={props.pagingPosition}
            rowClass={useCallback((value: any) => props.rowClass?.get(value)?.value ?? "", [props.rowClass])}
            setPage={setPage}
            setSortParameters={setSortParameters}
            settings={props.configurationAttribute}
            styles={props.style}
            valueForSort={useCallback(
                (value, columnIndex) => {
                    const column = props.columns[columnIndex];
                    return column.attribute ? column.attribute.get(value).value : "";
                },
                [props.columns]
            )}
            onSelect={selectActionProps.onSelect}
            onSelectAll={selectActionProps.onSelectAll}
            isSelected={selectActionProps.isSelected}
        />
    );
}

function transformColumnProps(props: ColumnsType[]): TableColumn[] {
    return props.map(prop => ({
        ...prop,
        header: prop.header && isAvailable(prop.header) ? prop.header.value ?? "" : "",
        sortable: isSortable(prop)
    }));
}
