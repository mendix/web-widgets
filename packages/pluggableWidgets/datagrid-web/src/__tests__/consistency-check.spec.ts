import { DatagridPreviewProps } from "../../typings/DatagridProps";
import { check } from "../Datagrid.editorConfig";

describe("consistency check", () => {
    describe("returns error when row select method and action trigger overlap", () => {
        test("selection: single by row click", () => {
            const props = {
                itemSelection: "Single",
                itemSelectionMethod: "rowClick",
                onClick: {},
                onClickTrigger: "single",
                columns: []
            };

            expect(check(props as unknown as DatagridPreviewProps)).toMatchSnapshot();
        });

        test("selection: multi by row click", () => {
            const props = {
                itemSelection: "Multi",
                itemSelectionMethod: "rowClick",
                onClick: {},
                onClickTrigger: "single",
                columns: []
            };

            expect(check(props as unknown as DatagridPreviewProps)).toMatchSnapshot();
        });
    });

    describe("returns empty array when row selection method and action trigger not overlap", () => {
        test("selection: none", () => {
            const props = {
                itemSelection: "None",
                itemSelectionMethod: "checkbox",
                onClick: {},
                onClickTrigger: "single",
                columns: []
            };

            expect(check(props as unknown as DatagridPreviewProps)).toEqual([]);
        });

        test("selection: single by checkbox", () => {
            const props = {
                itemSelection: "Single",
                itemSelectionMethod: "checkbox",
                onClick: {},
                onClickTrigger: "single",
                columns: []
            };

            expect(check(props as unknown as DatagridPreviewProps)).toEqual([]);
        });

        test("selection: single by checkbox, trigger: double", () => {
            const props = {
                itemSelection: "Single",
                itemSelectionMethod: "checkbox",
                onClick: {},
                onClickTrigger: "double",
                columns: []
            };

            expect(check(props as unknown as DatagridPreviewProps)).toEqual([]);
        });
    });
});
