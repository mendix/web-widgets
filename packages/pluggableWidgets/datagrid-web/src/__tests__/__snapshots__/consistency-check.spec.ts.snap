// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`consistency check returns error when row select method and action trigger overlap selection: multi by row click 1`] = `
[
  {
    "message": "The row click action is ambiguous. Change "On click trigger" to "Double click" or "Selection method" to "Checkbox".",
    "severity": "error",
  },
]
`;

exports[`consistency check returns error when row select method and action trigger overlap selection: single by row click 1`] = `
[
  {
    "message": "The row click action is ambiguous. Change "On click trigger" to "Double click" or "Selection method" to "Checkbox".",
    "severity": "error",
  },
]
`;
